File: packages/memory/src/producer.ts
/**
 * @fileoverview Memory producer implementation
 * @module @anyq/memory/producer
 */

import {
  BaseProducer,
  type PublishOptions,
  type HealthStatus,
  ConnectionError,
} from '@anyq/core';
import type { MemoryQueueConfig } from './config.js';
import { MemoryQueue } from './queue.js';
import { getQueue, registerQueue } from './registry.js';

/**
 * In-memory producer implementation
 *
 * Publishes messages to an in-memory queue for testing.
 *
 * @example
 * ```typescript
 * const producer = new MemoryProducer({
 *   driver: 'memory',
 *   queueName: 'test-queue',
 * });
 *
 * await producer.connect();
 * await producer.publish({ data: 'test' });
 * await producer.disconnect();
 * ```
 */
export class MemoryProducer<T = unknown> extends BaseProducer<T> {
  private queue: MemoryQueue<T> | null = null;
  private readonly queueName: string;

  constructor(config: MemoryQueueConfig) {
    super(config);
    this.queueName = config.queueName ?? 'default';
  }

  /**
   * Connect to the in-memory queue
   */
  async connect(): Promise<void> {
    if (this._connected) {
      return;
    }

    try {
      // Get or create the queue
      let queue = getQueue<T>(this.queueName);
      if (!queue) {
        const config = this.config as MemoryQueueConfig;
        queue = new MemoryQueue<T>(this.queueName, {
          maxMessages: config.maxMessages,
          maxAgeMs: config.maxAgeMs,
        });
        registerQueue(this.queueName, queue);
      }

      this.queue = queue;
      this._connected = true;
      this.logger.info('Memory producer connected', { queue: this.queueName });
    } catch (error) {
      throw new ConnectionError(
        'Failed to connect to memory queue',
        error instanceof Error ? error : undefined
      );
    }
  }

  /**
   * Disconnect from the queue
   */
  async disconnect(): Promise<void> {
    if (!this._connected) {
      return;
    }

    this.queue = null;
    this._connected = false;
    this.logger.info('Memory producer disconnected');
  }

  /**
   * Publish a single message
   */
  async publish(body: T, options?: PublishOptions): Promise<string> {
    if (!this.queue) {
      throw new ConnectionError('Producer not connected');
    }

    const messageId = this.queue.enqueue(body, {
      key: options?.key,
      headers: options?.headers,
    });

    this.logger.debug('Message published', {
      messageId,
      queue: this.queueName,
      key: options?.key,
    });

    return messageId;
  }

  /**
   * Publish multiple messages
   */
  async publishBatch(
    messages: Array<{ body: T; options?: PublishOptions }>
  ): Promise<string[]> {
    if (!this.queue) {
      throw new ConnectionError('Producer not connected');
    }

    const messageIds: string[] = [];

    for (const { body, options } of messages) {
      const id = this.queue.enqueue(body, {
        key: options?.key,
        headers: options?.headers,
      });
      messageIds.push(id);
    }

    this.logger.debug('Batch published', {
      count: messageIds.length,
      queue: this.queueName,
    });

    return messageIds;
  }

  /**
   * Get health status
   */
  async healthCheck(): Promise<HealthStatus> {
    return {
      healthy: this._connected,
      connected: this._connected,
      details: {
        queue: this.queueName,
        queueSize: this.queue?.size() ?? 0,
      },
    };
  }

  /**
   * Get the underlying queue (for testing)
   */
  getQueue(): MemoryQueue<T> | null {
    return this.queue;
  }
}

File: go/memory/producer.go
package memory

import (
	"context"

	"github.com/sns45/anyq/go/core"
)

// Producer publishes messages to an in-memory queue.
type Producer struct {
	core.BaseProducer
	queueName string
	cfg       Config
	queue     *Queue
}

// NewProducer builds a memory producer.
func NewProducer(cfg Config) *Producer {
	if cfg.Driver == "" {
		cfg.Driver = core.DriverMemory
	}
	p := &Producer{queueName: cfg.queueName(), cfg: cfg}
	p.InitBase(cfg.BaseQueueConfig)
	return p
}

// Connect attaches to (or creates) the named queue.
func (p *Producer) Connect(ctx context.Context) error {
	if p.IsConnected() {
		return nil
	}
	p.queue = GetOrCreateQueue(p.queueName, p.cfg.MaxMessages, p.cfg.MaxAgeMs)
	p.SetConnected(true)
	p.Logger.Info("memory producer connected", map[string]any{"queue": p.queueName})
	return nil
}

// Disconnect detaches from the queue.
func (p *Producer) Disconnect(ctx context.Context) error {
	if !p.IsConnected() {
		return nil
	}
	p.queue = nil
	p.SetConnected(false)
	p.Logger.Info("memory producer disconnected", nil)
	return nil
}

// Publish enqueues a single message.
func (p *Producer) Publish(ctx context.Context, body []byte, opts *core.PublishOptions) (string, error) {
	if p.queue == nil {
		return "", core.NewConnectionError("producer not connected", nil)
	}
	var key string
	var headers core.MessageHeaders
	if opts != nil {
		key = opts.Key
		headers = opts.Headers
	}
	id := p.queue.Enqueue(body, key, headers)
	p.Logger.Debug("message published", map[string]any{"messageId": id, "queue": p.queueName})
	return id, nil
}

// PublishBatch enqueues several messages.
func (p *Producer) PublishBatch(ctx context.Context, items []core.BatchItem) ([]string, error) {
	if p.queue == nil {
		return nil, core.NewConnectionError("producer not connected", nil)
	}
	ids := make([]string, 0, len(items))
	for _, item := range items {
		var key string
		var headers core.MessageHeaders
		if item.Options != nil {
			key = item.Options.Key
			headers = item.Options.Headers
		}
		ids = append(ids, p.queue.Enqueue(item.Body, key, headers))
	}
	p.Logger.Debug("batch published", map[string]any{"count": len(ids), "queue": p.queueName})
	return ids, nil
}

// HealthCheck reports producer health.
func (p *Producer) HealthCheck(ctx context.Context) (core.HealthStatus, error) {
	connected := p.IsConnected()
	details := map[string]any{"queue": p.queueName}
	if p.queue != nil {
		details["queueSize"] = p.queue.Size()
	}
	return core.HealthStatus{Healthy: connected, Connected: connected, Details: details}, nil
}

// Queue returns the underlying queue (for testing).
func (p *Producer) Queue() *Queue { return p.queue }

var _ core.Producer = (*Producer)(nil)
